csharp/src/Apache.Arrow/Arrays/ArrayDataConcatenator.cs

```diff
@@ -367,7 +367,7 @@ private ArrowBuffer ConcatenateUnionTypeBuffer()
 
                 foreach (ArrayData arrayData in _arrayDataList)
                 {
-                    builder.Append(arrayData.Buffers[0]);
+                    builder.Append(arrayData.Buffers[0].Span.Slice(arrayData.Offset, arrayData.Length));
```

Comment:
I'm using the offset here but I haven't tested this works fully with sliced arrays, I'll follow up on this in #41164.